meeko/molsetup.py

```diff
@@ -4,623 +4,451 @@
-        for neigh in self.get_neigh(idx):
-            if neigh in exclude:
-                continue
-            collected.append(neigh)
-            exclude.append(neigh)
-            self.walk_recursive(neigh, collected, exclude)
-        return collected
-
-    def copy_attributes_from(self, template):
-        """ copy attributes to duplicate the template setup
-        NOTE: the molecule will always keep the original setup (i.e., template)"""
-        # TODO enable some kind of plugin system here too, to allow other objects to
-        # add attributes?
-        # TODO although, this would make the setup more fragile-> better have attributes
-        # explicitely added here, and that's it
-        for attr in self.attributes_to_copy:
-            attr_copy = deepcopy(getattr(template, attr))
-            setattr(self, attr, attr_copy)
-        # TODO possible BUG? the molecule is shared by the different setups
-        #      if one of them alters the molecule, properties will not be the same
-        self.mol = template.mol
-
-    def merge_terminal_atoms(self, indices):
-        """for merging hydrogens, but will merge any atom or pseudo atom
-            that is bonded to only one other atom"""
-
-        for index in indices:
-            if len(self.graph[index]) != 1:
-                msg = "Atempted to merge atom %d with %d neighbors. "
-                msg += "Only atoms with one neighbor can be merged."
-                msg = msg % (index + 1, len(self.graph[index]))
-                raise RuntimeError(msg)
-            neighbor_index = self.graph[index][0]
-            self.charge[neighbor_index] += self.get_charge(index)
-            self.charge[index] = 0.0
-            self.set_ignore(index, True)
-            
-    def init_atom(self):
-        """ iterate through molecule atoms and build the atoms table """
-        raise NotImplementedError("This method must be overloaded by inheriting class")
-
-    def perceive_rings(self, keep_chorded_rings, keep_equivalent_rings):
-        """ populate the rings and aromatic rings tableshe atoms table:
-        self.rings_aromatics : list
-            contains the list of ring_id items that are aromatic
-        self.rings: dict
-            store information about rings, like if they have corners that can
-            be flipped and the graph of atoms that belong to them:
-
-                self.rings[ring_id] = {
-                                'corner_flip': False
-                                'graph': {}
-                                }
-
-            The atom is built using the `walk_recursive` method
-        self.ring_atom_to_ring_id: dict
-            mapping of each atom belonginig to the ring: atom_idx -> ring_id
+        NOT MODIFIED YET
         """
-
-        def isRingAromatic(ring_atom_indices):
-            for atom_idx1, atom_idx2 in self.get_bonds_in_ring(ring_atom_indices):
-                bond = self.mol.GetBondBetweenAtoms(atom_idx1, atom_idx2)
```

Comment:
This seems like it's an RDKit specific ring thing, so to what extent should I have rings implemented in MoleculeSetups vs in RDKitMoleculeSetups/the external tool interface?

Reply:
Absolutely, we can get rid of rings. And of other stuff:

`rings_aromatic` should be removed because FlexMacrocycle only needs to know whether or not each bond is rotatable. Aromatic bonds can't be broken anyway because they are set as non-rotatable. In the hypothetical scenario in which an aromatic bond is deliberately set as rotatable, then I'd argue that FlexMacrocycle should consider it as breakable.

`rings` should be removed because `FlexMacrocycle.collect_rings()` can invoke enumeration of the rings, no need to store the rings in an attribute of molsetup.

`is_chiral` was only needed to discourage FlexMacrocycle from breaking bonds in which one or both atoms are chiral, but this is only needed for the original method (https://pubs.acs.org/doi/full/10.1021/ci700036j) that did not use pseudo-atoms to preserve the geometry of the bond. With the current method, it is OK to break bonds involving chiral atoms.

`bond_order` is only ever used as a proxy for `is_rotatable` in convoluted ways. We are better off setting `is_rotatable` directly. The one exception is FlexMacrocycle being able to break double bonds, but this is questionable because we can't properly restore the bond geometry without a dihedral potential, so we probably better off breaking bonds that are rotatable